File: src/main/java/main/main/Robertify.java
package main.main;

import com.google.common.collect.Lists;
import com.google.common.util.concurrent.ThreadFactoryBuilder;
import io.sentry.Sentry;
import lavalink.client.io.LavalinkLoadBalancer;
import lavalink.client.io.jda.JdaLavalink;
import lombok.Getter;
import lombok.SneakyThrows;
import main.audiohandlers.GuildMusicManager;
import main.audiohandlers.RobertifyAudioManager;
import main.commands.contextcommands.ContextCommandManager;
import main.commands.prefixcommands.audio.SkipCommand;
import main.commands.prefixcommands.dev.test.MenuPaginationTestCommand;
import main.commands.prefixcommands.util.reports.ReportsEvents;
import main.commands.slashcommands.SlashCommandManager;
import main.commands.slashcommands.commands.management.dedicatedchannel.DedicatedChannelEvents;
import main.commands.slashcommands.commands.misc.poll.PollEvents;
import main.constants.ENV;
import main.events.LogChannelEvents;
import main.events.SuggestionCategoryDeletionEvents;
import main.events.VoiceChannelEvents;
import main.utils.EventWaiter;
import main.utils.apis.robertify.RobertifyAPI;
import main.utils.component.interactions.AbstractSlashCommand;
import main.utils.database.mongodb.AbstractMongoDatabase;
import main.utils.database.mongodb.cache.redis.GuildRedisCache;
import main.utils.json.AbstractJSONFile;
import main.utils.pagination.PaginationEvents;
import main.utils.votes.api.discordbotlist.DBLApi;
import me.duncte123.botcommons.web.WebUtils;
import net.dv8tion.jda.api.GatewayEncoding;
import net.dv8tion.jda.api.entities.Activity;
import net.dv8tion.jda.api.requests.GatewayIntent;
import net.dv8tion.jda.api.sharding.DefaultShardManagerBuilder;
import net.dv8tion.jda.api.sharding.ShardManager;
import net.dv8tion.jda.api.utils.ChunkingFilter;
import net.dv8tion.jda.api.utils.MemberCachePolicy;
import net.dv8tion.jda.api.utils.SessionControllerAdapter;
import net.dv8tion.jda.api.utils.cache.CacheFlag;
import org.discordbots.api.client.DiscordBotListAPI;
import org.jetbrains.annotations.NotNull;
import org.quartz.Scheduler;
import org.quartz.SchedulerException;
import org.quartz.impl.StdSchedulerFactory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Base64;

public class Robertify {

    private static final Logger logger = LoggerFactory.getLogger(Robertify.class);


    @Getter
    public static ShardManager shardManager;
    @Getter
    private static JdaLavalink lavalink;
    @Getter
    private static DiscordBotListAPI topGGAPI;
    @Getter
    private static DBLApi discordBotListAPI;
    @Getter
    private static RobertifyAPI robertifyAPI;
    @Getter
    private static final EventWaiter commandWaiter = new EventWaiter();
    @Getter
    private static Scheduler cronScheduler;

    static {
        try {
            Robertify.cronScheduler = new StdSchedulerFactory().getScheduler();
        } catch (SchedulerException e) {
            logger.error("I was unable to initialize the CRON scheduler!");
        }
    }

    public static void main(String[] args) {
        WebUtils.setUserAgent("Mozilla/Robertify / bombies#4445");

        try {
            lavalink = new JdaLavalink(
                    getIdFromToken(Config.getBotToken()),
                    Config.getShardCount(),
                    shardId -> Robertify.getShardManager().getShardById(shardId)
            );

            for (var node : Config.getLavaNodes())
                lavalink.addNode(node.getURI(), node.getPassword());

            lavalink.getLoadBalancer().addPenalty(LavalinkLoadBalancer.Penalties::getPlayerPenalty);
            lavalink.getLoadBalancer().addPenalty(LavalinkLoadBalancer.Penalties::getCpuPenalty);

            var thread = new ThreadFactoryBuilder().setNameFormat("RobertifyShutdownHook").build();
            Runtime.getRuntime().addShutdownHook(thread.newThread(() -> {
                logger.info("Destroying all players (If any left)");
                shardManager.getGuildCache().stream()
                        .filter(guild -> guild.getSelfMember().getVoiceState().inAudioChannel())
                        .forEach(guild -> {
                            GuildMusicManager musicManager = RobertifyAudioManager.getInstance().getMusicManager(guild);
                            musicManager.getScheduler().disconnect(false);
                        });
                logger.info("Killing cron scheduler");
                try {
                    getCronScheduler().clear();
                } catch (SchedulerException e) {
                    logger.error("I couldn't clear scheduling data!");
                }
                shardManager.shutdown();
            }));

            DefaultShardManagerBuilder jdaBuilder = DefaultShardManagerBuilder.createDefault(
                            Config.getBotToken(),
                            GatewayIntent.GUILD_VOICE_STATES
                    )
                    .setShardsTotal(Config.getShardCount())
                    .setBulkDeleteSplittingEnabled(false)
                    .setChunkingFilter(ChunkingFilter.NONE)
                    .setMemberCachePolicy(MemberCachePolicy.VOICE)

                    // Event Listeners
                    .addEventListeners(
                            lavalink,
                            VoiceChannelEvents.waiter,
                            commandWaiter,
                            new Listener(),
                            new VoiceChannelEvents(),
                            new DedicatedChannelEvents(),
                            new PollEvents(),
                            new SuggestionCategoryDeletionEvents(),
                            new ReportsEvents(),
                            new LogChannelEvents(),
                            new SkipCommand()
                    )
                    .setVoiceDispatchInterceptor(lavalink.getVoiceInterceptor())

                    // Test Listeners
                    .addEventListeners(
                            new MenuPaginationTestCommand()
                    )

                    // Button Listeners
                    .addEventListeners(
                            new PaginationEvents()
                    )

                    .enableCache(
                            CacheFlag.VOICE_STATE
                    )
                    .disableCache(
                            CacheFlag.ACTIVITY,
                            CacheFlag.EMOJI,
                            CacheFlag.CLIENT_STATUS,
                            CacheFlag.ROLE_TAGS,
                            CacheFlag.ONLINE_STATUS,
                            CacheFlag.STICKER,
                            CacheFlag.SCHEDULED_EVENTS
                    )
                    .setSessionController(new SessionControllerAdapter() {
                        @NotNull
                        @Override
                        public String getGateway() {
                            return Config.hasGatewayUrl() ? Config.getGatewayUrl() : super.getGateway();
                        }

                        @Override
                        public void setGlobalRatelimit(long ratelimit) {
                            if (Config.hasGatewayUrl())
                                super.setGlobalRatelimit(0);
                            else super.setGlobalRatelimit(ratelimit);
                        }

                        @Override
                        public long getGlobalRatelimit() {
                            if (Config.hasGatewayUrl())
                                return 0L;
                            return super.getGlobalRatelimit();
                        }
                    })
                    .setActivity(Activity.listening("Starting up..."));

            final var disabledIntents = Lists.newArrayList(
                    GatewayIntent.DIRECT_MESSAGE_TYPING,
                    GatewayIntent.GUILD_BANS,
                    GatewayIntent.GUILD_INVITES,
                    GatewayIntent.GUILD_MEMBERS,
                    GatewayIntent.GUILD_MESSAGE_TYPING,
                    GatewayIntent.GUILD_PRESENCES,
                    GatewayIntent.DIRECT_MESSAGE_REACTIONS,
                    GatewayIntent.SCHEDULED_EVENTS
            );

            final var enabledIntents = Lists.newArrayList(GatewayIntent.GUILD_MESSAGES);

            if (Config.getBoolean(ENV.MESSAGE_CONTENT_INTENT_ENABLED))
                enabledIntents.add(GatewayIntent.MESSAGE_CONTENT);
            else
                disabledIntents.add(GatewayIntent.MESSAGE_CONTENT);

            jdaBuilder.enableIntents(enabledIntents);
            jdaBuilder.disableIntents(disabledIntents);

            // Register all slash commands
            SlashCommandManager slashCommandManager = new SlashCommandManager();
            for (var cmd : slashCommandManager.getGlobalCommands()) {
                jdaBuilder.addEventListeners(cmd);
                logger.debug("Registered the \"{}\" command.", cmd.getName());
            }
            for (var cmd : slashCommandManager.getGuildCommands()) {
                jdaBuilder.addEventListeners(cmd);
                logger.debug("Registered the \"{}\" command.", cmd.getName());
            }
            for (var cmd : slashCommandManager.getDevCommands()) {
                jdaBuilder.addEventListeners(cmd);
                logger.debug("Registered the \"{}\" command.", cmd.getName());
            }

            ContextCommandManager contextCommandManager = new ContextCommandManager();
            for (var cmd : contextCommandManager.getCommands()) {
                jdaBuilder.addEventListeners(cmd);
                logger.debug("Registered the \"{}\" context command.", cmd.getName());
            }

            // Initialize the JSON directory
            // This is a deprecated feature and is marked for removal
            // Until everything is fully removed, this method needs to be enabled
            // For a proper first-boot.
            AbstractJSONFile.initDirectory();

            AbstractMongoDatabase.initAllCaches();
            logger.info("Initialized all caches");

            GuildRedisCache.getInstance().loadAllGuilds();
            logger.info("All guilds have been loaded into cache");

            shardManager = jdaBuilder.build();

            if (Config.loadCommands())
                AbstractSlashCommand.loadAllCommands();

            Robertify.cronScheduler.start();

            initVoteSiteAPIs();

            if (Config.hasValue(ENV.ROBERTIFY_API_PASSWORD))
                    robertifyAPI = RobertifyAPI.ins;

            if (Config.hasValue(ENV.SENTRY_DSN))
                Sentry.init(options -> {
                    options.setDsn(Config.get(ENV.SENTRY_DSN));
                    options.setTracesSampleRate(1.0);
                });
        } catch (Exception e) {
            logger.error("[FATAL ERROR] An unexpected error occurred!", e);
        }
    }

    private static void loadNeededGlobalCommands() {

    }

    private void loadGlobalSlashCommands() {
        AbstractSlashCommand.loadAllCommands();
    }

    public static void initVoteSiteAPIs() {
        if (!Config.get(ENV.TOP_GG_TOKEN).isEmpty())
            topGGAPI = new DiscordBotListAPI.Builder()
                    .token(Config.get(ENV.TOP_GG_TOKEN))
                    .botId(getIdFromToken(Config.get(ENV.BOT_TOKEN)))
                    .build();

        if (!Config.get(ENV.DBL_TOKEN).isEmpty())
            discordBotListAPI = new DBLApi.Builder()
                    .setToken(Config.get(ENV.DBL_TOKEN))
                    .setBotID(getIdFromToken(Config.get(ENV.BOT_TOKEN)))
                    .build();
    }

    private static String getIdFromToken(String token) {
        return new String(
                Base64.getDecoder().decode(
                        token.split("\\.")[0]
                )
        );
    }
}

File: src/main/kotlin/main/main/Robertify.kt
package main.main

//import main.utils.resume.GuildResumeManager
import api.RobertifyKtorApi
import com.adamratzman.spotify.SpotifyAppApi
import com.adamratzman.spotify.spotifyAppApi
import com.google.common.util.concurrent.ThreadFactoryBuilder
import dev.arbjerg.lavalink.client.*
import dev.arbjerg.lavalink.client.event.TrackEndEvent
import dev.arbjerg.lavalink.client.event.TrackExceptionEvent
import dev.arbjerg.lavalink.client.event.TrackStartEvent
import dev.arbjerg.lavalink.client.event.TrackStuckEvent
import dev.arbjerg.lavalink.client.loadbalancing.RegionGroup
import dev.arbjerg.lavalink.libraries.jda.JDAVoiceUpdateListener
import dev.minn.jda.ktx.events.listener
import dev.minn.jda.ktx.jdabuilder.injectKTX
import dev.minn.jda.ktx.util.SLF4J
import io.sentry.Sentry
import kotlinx.coroutines.runBlocking
import main.audiohandlers.RobertifyAudioManager
import main.commands.slashcommands.SlashCommandManager
import main.commands.slashcommands.SlashCommandManager.registerCommands
import main.constants.ENV
import main.events.EventManager
import main.main.Listener.Companion.loadNeededSlashCommands
import main.main.Listener.Companion.rescheduleUnbans
import main.utils.api.robertify.RobertifyApi
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.database.mongodb.AbstractMongoDatabase
import main.utils.database.mongodb.cache.BotDBCache
import main.utils.json.reminders.RemindersConfig
import net.dv8tion.jda.api.OnlineStatus
import net.dv8tion.jda.api.entities.Activity
import net.dv8tion.jda.api.events.guild.GuildReadyEvent
import net.dv8tion.jda.api.events.session.ReadyEvent
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.sharding.DefaultShardManagerBuilder
import net.dv8tion.jda.api.sharding.ShardManager
import net.dv8tion.jda.api.utils.MemberCachePolicy
import net.dv8tion.jda.api.utils.cache.CacheFlag
import org.discordbots.api.client.DiscordBotListAPI
import org.quartz.SchedulerException
import org.quartz.impl.StdSchedulerFactory
import java.util.*

object Robertify {
    private val logger by SLF4J
    val cronScheduler = StdSchedulerFactory.getDefaultScheduler()
    var topGGAPI: DiscordBotListAPI? = null
    lateinit var lavalink: LavalinkClient
        private set
    lateinit var shardManager: ShardManager
        private set
    lateinit var spotifyApi: SpotifyAppApi
        private set
    lateinit var externalApi: RobertifyApi
        private set


    fun main() = runBlocking {
        if (Config.hasValue(ENV.SENTRY_DSN))
            Sentry.init { options ->
                options.dsn = Config.SENTRY_DSN
                options.environment = Config.ENVIRONMENT
            }

        // Setup graceful shutdown hooks
        val mainShutdownHook = ThreadFactoryBuilder()
            .setNameFormat("RobertifyShutdownHook")
            .build()
        Runtime.getRuntime().addShutdownHook(mainShutdownHook.newThread {
            logger.info("Destroying all players (If any left)")
            runBlocking {
                shardManager.guilds
                    .filter { guild -> guild.selfMember.voiceState != null }
                    .forEach { guild -> guild.jda.directAudioController.disconnect(guild) }
            }
        })

        val cronShutdownHook = ThreadFactoryBuilder()
            .setNameFormat("RobertifyCronShutdownHook")
            .build()
        Runtime.getRuntime().addShutdownHook(cronShutdownHook.newThread {
            logger.info("Killing cron scheduler")
            try {
                cronScheduler.clear()
            } catch (e: SchedulerException) {
                logger.error("I couldn't clear scheduling data!")
            }
        })

        // Init caches
        AbstractMongoDatabase.initAllCaches()
        logger.info("Initialized all caches.")


        // Setup LavaLink
        logger.info("Setting up LavaLink...")
        lavalink = LavalinkClient(getIdFromToken(Config.BOT_TOKEN).toLong())
        lavalink.on<dev.arbjerg.lavalink.client.event.ReadyEvent>()
            .subscribe { event ->
                logger.info("LavaLink ready on node ${event.node.name}. Session ID is ${event.sessionId}")
            }
        logger.debug("LavaLink ReadyEvent registered.")

        Config.LAVA_NODES.forEach { node ->
            val lavaNode = lavalink.addNode(
                NodeOptions.Builder(
                    serverUri = node.uri,
                    password = node.password,
                    name = node.name,
                    regionFilter = RegionGroup.US
                ).build()
            )

            lavaNode.on<TrackStartEvent>().subscribe { event ->
                val musicManager = RobertifyAudioManager.getMusicManager(event.guildId)
                musicManager.scheduler.onTrackStart(event)
            }
            lavaNode.on<TrackEndEvent>().subscribe { event ->
                val musicManager = RobertifyAudioManager.getMusicManager(event.guildId)
                musicManager.scheduler.onTrackEnd(event)
            }
            lavaNode.on<TrackStuckEvent>().subscribe { event ->
                val musicManager = RobertifyAudioManager.getMusicManager(event.guildId)
                musicManager.scheduler.onTrackStuck(event)
            }
            lavaNode.on<TrackExceptionEvent>().subscribe { event ->
                val musicManager = RobertifyAudioManager.getMusicManager(event.guildId)
                musicManager.scheduler.onTrackException(event)
            }

            logger.info("Registered lava node with address: ${node.uri}")
        }

        // Build bot connection
        logger.info("Building shard manager...")
        val shardManagerBuilder = DefaultShardManagerBuilder.createLight(
            Config.BOT_TOKEN,
            listOf(GatewayIntent.GUILD_VOICE_STATES, GatewayIntent.GUILD_MESSAGE_REACTIONS),
        )
            .apply {
                setShardsTotal(Config.SHARD_COUNT)
                injectKTX()
                setMemberCachePolicy(MemberCachePolicy.DEFAULT)
                setBulkDeleteSplittingEnabled(false)
                enableCache(CacheFlag.VOICE_STATE)
                setVoiceDispatchInterceptor(JDAVoiceUpdateListener(lavalink))
                setActivity(Activity.listening("Starting up..."))

                val disabledIntents = mutableListOf(
                    GatewayIntent.DIRECT_MESSAGE_TYPING,
                    GatewayIntent.GUILD_MODERATION,
                    GatewayIntent.GUILD_INVITES,
                    GatewayIntent.GUILD_MEMBERS,
                    GatewayIntent.GUILD_MESSAGE_TYPING,
                    GatewayIntent.GUILD_PRESENCES,
                    GatewayIntent.DIRECT_MESSAGE_REACTIONS,
                    GatewayIntent.SCHEDULED_EVENTS
                )

                val enabledIntents = mutableListOf(GatewayIntent.GUILD_MESSAGES)

                if (Config.MESSAGE_CONTENT_ENABLED)
                    enabledIntents.add(GatewayIntent.MESSAGE_CONTENT)
                else disabledIntents.add(GatewayIntent.MESSAGE_CONTENT)

                enableIntents(enabledIntents)
                disableIntents(disabledIntents)
            }

        shardManager = shardManagerBuilder.build()
        logger.info("Successfully built shard manager")

        shardManager.handleShardReady()
        shardManager.handleGuildReady()

        val slashCommandManager = SlashCommandManager
        shardManager.registerCommands(
            slashCommandManager.guildCommands
                .merge(slashCommandManager.globalCommands, slashCommandManager.devCommands)
        )
        logger.info("Registered all slash commands.")

        EventManager.registeredEvents
        logger.info("Registered all event controllers.")

        if (Config.LOAD_COMMANDS)
            AbstractSlashCommand.loadAllCommands()

        if (Config.LOAD_NEEDED_COMMANDS)
            loadNeededGlobalCommands()

        cronScheduler.start()
        initVoteSiteAPIs()

        // Setup SpotifyAPI
        spotifyApi = spotifyAppApi(
            clientId = Config.SPOTIFY_CLIENT_ID,
            clientSecret = Config.SPOTIFY_CLIENT_SECRET
        ).build(true)

        if (Config.hasValue(ENV.ROBERTIFY_API_PASSWORD))
            externalApi = RobertifyApi()

        RobertifyAudioManager.scheduleCleanup()
        RobertifyKtorApi.start()
    }

    private fun ShardManager.handleShardReady() = listener<ReadyEvent> { event ->
        val jda = event.jda
        logger.info("Watching ${event.guildAvailableCount} guilds on shard #${jda.shardInfo.shardId} (${event.guildUnavailableCount} unavailable)")
        BotDBCache.instance.lastStartup = System.currentTimeMillis()
        jda.shardManager?.setPresence(OnlineStatus.ONLINE, Activity.listening("/help"))
    }

    private fun ShardManager.handleGuildReady() = listener<GuildReadyEvent> { event ->
        val guild = event.guild
        loadNeededSlashCommands(guild)

        try {
            rescheduleUnbans(guild)
        } catch (e: Exception) {
            logger.error("Failed to reschedule unbans for guild ${guild.id}", e)
        }

        try {
            RemindersConfig(guild).scheduleReminders()
        } catch (e: Exception) {
            logger.error("Failed to schedule reminders for guild ${guild.id}", e)
        }
//        GuildResumeManager(guild).loadTracks()
    }

    fun initVoteSiteAPIs() {
        if (Config.TOP_GG_TOKEN.isNotEmpty())
            topGGAPI = DiscordBotListAPI.Builder()
                .token(Config.TOP_GG_TOKEN)
                .botId(getIdFromToken(Config.BOT_TOKEN))
                .build()
    }

    private fun getIdFromToken(token: String) =
        String(Base64.getDecoder().decode(token.split("\\.".toRegex())[0]))

    private fun List<AbstractSlashCommand>.merge(vararg lists: List<AbstractSlashCommand>): List<AbstractSlashCommand> =
        this + lists.reduce { acc, next -> acc + next }

    private fun loadNeededGlobalCommands() {

    }
}

fun main(args: Array<String>) = Robertify.main()